src/typeahead/typeahead-window.ts

```diff
@@ -116,6 +184,8 @@ export class NgbTypeaheadWindow implements OnInit {
 
   select(item) { this.selectEvent.emit(item); }
 
+  getContext() { return this; }
```

Comment:
I am not keen of opening this kind of hole. Not for public API.

Reply:
It is not the "public" API, but it is required to pass the 'window' to the 'window template' context and then to the 'window item' as an input.
It is a public method, but not a public API (same as `.select()` method).

https://github.com/ng-bootstrap/ng-bootstrap/pull/3711/files#diff-2b05df9eab924379f039993fae7b433eR101

That is the biggest bottleneck in the PR, but I've no idea how to solve it...

Reply:
My point here is to **not** return _this_ !! We return an object containing only what we need to access.

```typescript
getContext() {
return {
  markActive: ...
  select: ... 
  id: ...
  activeIdx: ...
  resultTemplate: ...
}
```

If we need them _live_ we could keep getters so they are executed when accessed, but we don't return _this_. To me, this is not safe/maintainable.

Reply:
Other solution, we could also pass every single needed property as one on the `templateContext` itself, instead of forwarding _this_

Then people would write
```html
<.  ..... let-markActive="markActive" let-select="select"  .... >
```

Reply:
I don't think there was any intention to give this kind of API for people. What's the point, if we handle all this internally?

Reply:
On the other hand I see what you mean. If we have this API, we could use it and then people could use it. And it might be cleaner overall.

Reply:
Let me rephrase my concern here:

👍 I am not against the surface API provided by an input `@Input() context` on the component.

👎 I am against providing `this` as a value to it. `this` will move over time, it's opaque, you don't know what it provides. Forcing us to declare what we export as context makes us design better API.

And finally, for pure documentation, we **should** use an interface to declare this context, to avoid such `any` effect as shown below

![image](https://user-images.githubusercontent.com/1152740/82227325-e734e880-9927-11ea-8217-f7c6cc9822c1.png)

Reply:
> 
> 
> Let me rephrase my concern here:
> 
> 👍 I am not against the surface API provided by an input `@Input() context` on the component.
> 
> 👎 I am against providing `this` as a value to it. `this` will move over time, it's opaque, you don't know what it provides. Forcing us to declare what we export as context makes us design better API.
> 
> And finally, for pure documentation, we **should** use an interface to declare this context, to avoid such `any` effect as shown below
> 
> ![image](https://user-images.githubusercontent.com/1152740/82227325-e734e880-9927-11ea-8217-f7c6cc9822c1.png)

I would like to make context not visible by users at all. We would like to change what fields belong to context instead of exposing them to users. The only problem is how to inject context in items without users noticing. It isn't meant to be used by users, that's why it has the generic "any". It should not be used by users at all.